smithy-model/src/main/java/software/amazon/smithy/model/knowledge/PropertyBindingIndex.java

```diff
@@ -0,0 +1,206 @@
+     * member shape does not correspond to a property.
+     *
+     * @param memberShapeId the ShapeId of the member shape to get the property name for.
+     * @return the property name for a given member shape if there is one.
+     */
+    public Optional<String> getPropertyName(ShapeId memberShapeId) {
+        return Optional.ofNullable(memberShapeToPropertyName.get(memberShapeId));
+    }
+
+    /**
+     * Returns true if a member shape positively maps to a property.
+     *
+     * {@see PropertyBindingIndex#getPropertyName(ShapeId)} will return a non-empty Optional if this method
+     * returns true.
+     *
+     * @param memberShapeId the ShapeId of the member shape to check
+     * @return true if member shape maps to a property on the given resource
+     */
+    public boolean isMemberShapeProperty(ShapeId memberShapeId) {
+        return memberShapeToPropertyName.containsKey(memberShapeId);
+    }
+
+    /**
+     * Resolves and returns the output shape of an operation that contains the
+     * top-level resource bound properties. Handles adjustments made with
+     * @nestedProperties trait.
+     *
+     * @param operation operation to retrieve output properties shape for.
+     * @return the output shape of an operation that contains top-level resource
+     *  properties.
+     */
+    public Shape getOutputPropertiesShape(OperationShape operation) {
+        Model model = getModel();
+        return getPropertiesShape(operationIndex.getOutputMembers(operation).values(),
+                model.expectShape(operation.getOutputShape()));
+    }
+
+    /**
+     * Resolves and returns the input shape of an operation that contains the
+     * top-level resource bound properties. Handles adjustments made with
+     * @nestedProperties trait.
+     *
+     * @param operation operation to retrieve output properties shape for
+     * @return the input shape of an operation that contains top-level resource
+     *  properties.
+     */
+    public Shape getInputPropertiesShape(OperationShape operation) {
+        Model model = getModel();
+        return getPropertiesShape(operationIndex.getInputMembers(operation).values(),
+                model.expectShape(operation.getInputShape()));
+    }
+
+    /**
+     * Returns true if member is required to have an associated property mapping.
+     *
+     * @param memberShapeId the ShapeId of the member shape to check
+     * @return True if input/output member is required to have a property mapping.
+     */
+    public boolean doesMemberShapeRequireProperty(ShapeId memberShapeId) {
+        return !memberShapeDoesNotRequireProperty.getOrDefault(memberShapeId, false);
+    }
+
+    private Model getModel() {
+        return Objects.requireNonNull(model.get(), "The dereferenced WeakReference<Model> is null");
+    }
+
+    private Set<ShapeId> computeNotPropertyTraits() {
+        Model model = getModel();
+        return model.shapes()
```

Comment:
This should really use `model.getShapesWithTrait(NotPropertyTrait.class)` to get the filtered initial set of shapes. Then filter it down further to find matching shapes that have the trait trait.